Add new cmdlet Get-PnPDiagnostics to get diagnostic information of current PnP PowerShell status

jackpoz opened the pull request:
## Type ##
- [ ] Bug Fix
- [x] New Feature
- [ ] Sample

## Related Issues? ##
Fixes #2719, follow up of #2728

## What is in this Pull Request ? ##
This PR is a draft and should not be merged yet. See the task list below after the screenshots.

This PR adds a new command Get-PnPDiagnostics that returns information useful for troubleshooting, both locally for users or when filing a new issue on GitHub.

Here are a few examples:
- Executed before any other command (not connected to any site)
![image](https://user-images.githubusercontent.com/1153754/85882239-34399380-b7df-11ea-9966-98151a66c7ca.png)

- Connected to a site
![image](https://user-images.githubusercontent.com/1153754/85882422-88447800-b7df-11ea-97e6-aaa4987cf3d1.png)

- After an exception
![image](https://user-images.githubusercontent.com/1153754/85889854-8c2ac700-b7ec-11ea-95c3-cfdb9dae6a3f.png)


Fields included (the flagged ones are implemented, the others require some guidance) :
- [x] Version (string): PnP PowerShell version, i.e.: "3.22.2006.2"
- [x] Platform (string): PnP PowerShell targeted platform, i.e.: "SPO". Possible values are "SP2013", "SP2016", "SP2019", "SPO", "ONPREMISES" (only for future SP20xx versions if this command is not updated)
- [x] ModuleName (string): This is currently defined in SharePointPnP.PowerShell.ModuleFilesGenerator project and I couldn't find a way to retrieve this from C# in the same way Get-Module returns it. If someone has some tips/advice, please let me know.
- [x] OperatingSystem (string): Operating System retrieved from .NET Framework API, i.e.: Microsoft Windows NT 10.0.18363.0
- [x] ConnectionMethod: Connection method of current connection if any. It's the enum SharePointPnP.PowerShell.Commands.Model.ConnectionMethod . I know @erwinvanhunen asked for "Authentication Mode" and I wonder if this gives the same information or if I picked the wrong property.
- [x] CurrentSite (string): Url of the current connection, if any.
- [x] AccessTokenExpirationTime: This is a bit tricky because AccessTokens are stored as Dictionary with TokenAudience (SharePointOnline, MicrosoftGraph, OfficeManagementApi) as key and I couldn't find a clear relation between the current connection and which TokenAudience is the active one (or if there can be more than one active at the same time). Should the command display the expiration time of any token found ?
- [x] ModulePath (string): I tried to look for a way to retrieve from C# information that is actually available only from PowerShell and I couldn't find any. If someone has some tips/advice, please let me know.
- [x] NewerVersionAvailable (bool): True if there is a newer version, False otherwise.
- [x] LastCorrelationId (string): Same value returned by Get-PnPException
- [x] LastExceptionTimeStampUtc (Nullable&lt;DateTime&gt;): Same value returned by Get-PnPException
- [x] LastExceptionMessage (string): Same value returned by Get-PnPException
- [x] LastExceptionStacktrace (string): Same value returned by Get-PnPException
- [x] LastExceptionScriptLineNumber (Nullable&lt;int&gt;): Same value returned by Get-PnPException

To discuss:
- [x] How to implement the missing fields
- [x] Does the output look ok as class or ? Is there a better way to display nested properties of objects, like List&lt;T&gt; or PnPException ?
- [ ] As Last Exception has a few fields, I ended up adding each field by itself. Does this look ok ?

Please let me know what to change (and maybe suggest how 😃 ) . "Allow edits by maintainers" is enabled if needed.

KoenZomers commented:
Haven't looked at the code behind this already, but from the end result, it looks nice!

> AccessTokenExpirationTime: This is a bit tricky because AccessTokens are stored as Dictionary with TokenAudience (SharePointOnline, MicrosoftGraph, OfficeManagementApi) as key and I couldn't find a clear relation between the current connection and which TokenAudience is the active one (or if there can be more than one active at the same time). Should the command display the expiration time of any token found ?

One connected instance can have 1 to many tokens. Currently we support three (SPO, Graph, O365 Management) but there could and will be more in the future. Would be super useful if you can include listing the active tokens. Something like a summary of Get-PnPGraphAccessToken and Get-PnPOffice365ManagementApiAccessToken (or something, written from my memory). There is no such cmdlet for the SharePoint token yet at the moment.

> ModulePath (string): I tried to look for a way to retrieve from C# information that is actually available only from PowerShell and I couldn't find any. If someone has some tips/advice, please let me know.

I would say you can use the path from which the module executes. Check one of the base cmdlets where we use this path to add a reference to newtonsoft.json. I think you can use the same.

Keep up the good work on this one!

KoenZomers commented:
To add:

>ConnectionMethod: Connection method of current connection if any. It's the enum SharePointPnP.PowerShell.Commands.Model.ConnectionMethod . I know @erwinvanhunen asked for "Authentication Mode" and I wonder if this gives the same information or if I picked the wrong property.

This is being kept, but likely not totally accurate as by now there are way more connection options that can be used than that are available in this property. During my massive auth rewrite I didn't dare to touch it yet as I didn't know where this value is actually used for something or if this would be purely informational only for the statistical logging. This would be something that needs looking into before we can make these reliable. For now, I would say, you can add it to your cmdlet already to output whatever it has been set to. We can then fix that whatever it has been set to in later PRs, if needed.

jackpoz commented:
I changed the output to a PSObject so new properties can easily be added (learning the PS SDK a bit at a time 😃 ) .

I implemented the AccessTokenExpirationTime{Token Audience name} as a property added only if there is a token. I preferred to enumerate the possible token values in Get-PnPDiagnostics so PnPConnection exposes only a TryGetTokenExpirationTime() method, instead of exposing the whole AccessTokens private container.

This is how it looks before and after calling Connet-PnPOnline with MS Graph
![image](https://user-images.githubusercontent.com/1153754/86167918-a8de3c00-bb17-11ea-9d35-4f24f7dc64e4.png)

You can see how "AccessTokenExpirationTimeMicrosoftGraph" property appears only in the second execution of Get-PnPDiagnostics. The reason for this choice was to easily support N number of tokens without making the output too long (speaking of which, let me know if this should apply to other fields too, adding them only if not null).

KoenZomers commented:
I would prefer always seeing the same parameters in the output and just leaving them empty/null if there's no value for it. Different outputs based on circumstances is very confusing and not how other cmdlets work.

I would prefer seeing a collection with tokens as there could be multiple tokens and new ones could be added in the future. So instead of specifically having a property AccessTokenExpirationTime*MicrosoftGraph* just having something like AccessTokens which is a generic collection of tokens under which there will be a type, exactly like it already is today in the PnP code.

jackpoz commented:
I implemented ModulePath, ModuleName and AccessTokens as read-only dictionary (I tried to expose the bare minimum required from PnPConnection type).

Output with no connection:
![image](https://user-images.githubusercontent.com/1153754/86496022-cdd2e900-bd7b-11ea-860f-1fe7eb1eb4c7.png)

Output with MS Graph connection:
![image](https://user-images.githubusercontent.com/1153754/86496112-33bf7080-bd7c-11ea-9b30-0a948caf6221.png)

The token itself can be retrieved either specifying the key like
```powershell
(Get-PnPDiagnostics).AccessTokens[([SharePointPnP.PowerShell.Commands.Enums.TokenAudience]::MicrosoftGraph)]
```
or listing all the available ones like
```powershell
(Get-PnPDiagnostics).AccessTokens.Values
```

@KoenZomers please let me know if a read-only dictionary is the desired collection type or if a List<GenericToken> is preferred.
I have to test ModulePath and ModuleName in an Azure Function, other than that I think this PR will be ready for review.

KoenZomers commented:
Looks really good to me! @erwinvanhunen is currently working on some changes in the area of working with the tokens. He mentioned to push his updates to the dev branch later this weekend. Let's see what those changes will be and if it will still work like this. If it does, I like this way of accessing the tokens!

Another remark, I would write the Platform output in full, so SharePointOnline, SharePoint2013, SharePoint 2016, SharePoint 2019. No abbreviations.

jackpoz commented:
I updated the platform (not sure if you wanted a space between SharePoint and the version, I picked the no-space one for now).
I rebased the PR on latest dev, I'll try to keep merging dev into this PR to spot any conflict (strangely github still reported conflicts after I merged).

jackpoz commented:
On Azure GetExecutingAssembly().Location would return the temporary folder used by the host to run the Azure Function, while GetExecutingAssembly().CodeBase returning the actual folder where I deployed the module, so I swapped the order to prefer CodeBase .
Locally they return the same result so it shouldn't cause any issue.

This is the output from Azure Function:
![image](https://user-images.githubusercontent.com/1153754/86534285-924d3180-bed7-11ea-9f46-eb0845b9f887.png)

erwinvanhunen commented:
Closing this PR. We have implemented it in pnp.powershell: https://github.com/pnp/powershell
